### ISSUE TITLE
DNSSD backend based connection issue after successful discovery

### ISSUE THREAD

**Issue by FueLee** [OpenPrinting/cups#1091]
**Before you write the report**
Read the [REPORTING_ISSUES.md](https://github.com/OpenPrinting/cups/blob/master/REPORTING_ISSUES.md) file in the main repository and prepare data mentioned there which looks relevant to you issue.

**Describe the bug**
In case the Service Name contain spaces, after the successful DNSSD backend based discovery, the DNSSD based connection to the printer fails.

**To Reproduce**
1. Configure a printer with spaces in the Service Name
2. Add the printer with CUPS WebUI to the system
3. After successful installation activate the Print Test Page function from CUPS WebUI (Printers/My Printer/Print Test Page)
4. Cannot connect to the printer, printing fails

**Expected behavior**
Based on the documentation of CUPS, the service name can contain spaces and after successful installation of such printer, CUPS can print test page.

- CUPS documentation URL: http://localhost:631/help/network.html?TOPIC=Getting+Started&QUERY=
- Example URI: dnssd://Foo%20Fighter-1969._pdl-datastream._tcp.local./

**Screenshots**
Not applicable

**System Information:**
 - OS and its version: Ubuntu 24.10
 - CUPS version: 2.4.10
 - Printer: EPSON L3150
 - Supported protocols by the printer: LPD, JetDirect, Bonjour, etc.
 - Printer driver is provided by vendor (Epson)

**Additional context**
- After a lot of failed tries with DNSSD, I could configure CUPS to use LPD (lpd://) and/or JetDirect (socket://) directly, and the printer/printing is working.
- After I removed the spaces from the Service Name field (in the printer administration webpage), I could also print the test page with DNSSD configuration.
- When the Service Name contain spaces:
  - the ```avahi-browse -art``` command failed to resolve the services.
  - the generated URI is: dnssd://Epson%20printing%20services._pdl-datastream._tcp.local/
- When the Service Name does not contain spaces:
  - the ```avahi-browse -art``` command successfully resolve the services.
  - the generated URI is: dnssd://epson._pdl-datastream._tcp.local/

I configured my printer long time ago and it was working well with all kind of non Linux devices. So, the long story short, I was happy to see that the installation process successfully recognized my printer with DNSSD, but right after the successful installation of driver, I could not use it (and the documentation of CUPS encouraged me to rule out the "spaces in the service name" as a root cause).

**Comment by dendemann** [comment]
Maybe I just ran into the same issue. Never had printing issues before. After switching to LPD protocol I got it printing again. 

#### System Information:

* OS: Arch linux (6.6.59-1-lts)
* CUPS version: 2.4.11
* Printer: Brother HL-L8250CDN
* Printer driver is provided by vendor and made available via AUR package

```
lpinfo -v
network beh
file cups-pdf:/
network socket
network https
network http
network ipps
network ipp
network lpd
network smb
network dnssd://Brother%20HL-L8250CDN%20series._pdl-datastream._tcp.local/?uuid=e32...
network ipp://Brother%20HL-L8250CDN%20series._ipp._tcp.local/
network lpd://BRN3005.../BINARY_P1
```

**Comment by zdohnal** [comment]
Please provide debug2 logs from CUPS when you try to print - we have to ensure we send the correct hostname to Avahi, otherwise if we do, it will be Avahi issue, not CUPS.

I cannot reproduce it in my env - test page prints fine:

Fedora 40
CUPS 2.4.11
Service Name: "HP Color LaserJet MFP M277dw"

Since it works for me here it looks like distro specific/environment specific as well, related to mDNS resolution.

But please attach .txt with debug2 logs, to be sure we do the right with on CUPS side.

**Comment by FueLee** [comment]
The log files are attached.
[horatius_access_log.txt](https://github.com/user-attachments/files/17660081/horatius_access_log.txt)
[horatius_error_log.txt](https://github.com/user-attachments/files/17660082/horatius_error_log.txt)

Since logrotate deleted my previous logs, I reconfigured my printer to have spaces in the service name (and renamed it to "Horatius printing services") and also did the below steps to reproduce:

1. Printer reconfiguration, service name is "Horatius printing services" now, restart printer.
2. On CUPS interface "Find new printers" action that discovered the printer with the new name. The URI is: "dnssd://Horatius%20printing%20services._pdl-datastream._tcp.local/"
3. sudo cupsctl --debug-logging
4. systemctl restart cups around 11:15
5. On CUPS interface from Maintenance, "Print Test Page" around 11:17
6. The error message in CUPS interface: 
```bash
Final error message:
processing since
Thu 07 Nov 2024 11:17:02 AM CET 
"Unable to locate printer."
```

>Please note that in the log files you might find "horatius" or "horatius-mDNS" references for my old printer configuration for DNSSD (without spaces in the service name) that is a working configuration for me. I did not deleted this printer configuration for this test, but might be confusing...

Thanks

**Comment by FueLee** [comment]
Ok, I found the root cause and it is neither CUPS nor Avahi (or distro). It is the printer.

I ran several tests and once the test page was printed, so I ran the ```avahi-browse -art``` command and got back the successful name resolution for service name with spaces. But few minutes later it failed again...

- So I restarted the printer and checked again with the avahi command and was successful.
- Waited a minute and checked again, it was successful
- Waited again minute or so and checked again, it was successful
- Waited again minute or so and checked again and it failed!

The time difference was about 4 minutes. It seems that this printer shutdown (or crash?) its internal avahi service, because I could reproduce this behavior few more times.

Of course, the LPD and JetDirect printing method were working, so it is not the energy saving settings of the printer (which was set to 1 hour btw).

I am sorry for the troubles, but the mystery is solved in my case.
